### PR TITLE
[SPARK-7277][SQL] Throw exception if the property mapred.reduce.tasks is set to -1

### PR DESCRIPTION
**Pull request by viirya**
JIRA: https://issues.apache.org/jira/browse/SPARK-7277

As automatically determining the number of reducers is not supported (`mapred.reduce.tasks` is set to `-1`), we should throw exception to users.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/commands.scala [hunk 1]
```diff
@@ -84,8 +84,15 @@ case class SetCommand(
       logWarning(
         s"Property ${SQLConf.Deprecated.MAPRED_REDUCE_TASKS} is deprecated, " +
           s"automatically converted to ${SQLConf.SHUFFLE_PARTITIONS} instead.")
-      sqlContext.setConf(SQLConf.SHUFFLE_PARTITIONS, value)
-      Seq(Row(s"${SQLConf.SHUFFLE_PARTITIONS}=$value"))
+      if (value == "-1") {
+        logWarning(
+          s"Set this property to -1 for automatically determining the number of reducers " +
+            s"is not supported, showing current ${SQLConf.SHUFFLE_PARTITIONS} instead.")
```
Comment (squito) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/commands.scala]: typo: "Setting this property ..."

Comment (squito) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/commands.scala]: actually can I suggest a small rewording a think is a little clearer?

`Setting this property to -1 for automatically determing the number of reducers is not supported.  Ignoring request, and showing  ${SQLConf.SHUFFLE_PARTITIONS} instead`

**Comment by squito** [comment]
could you also add a test that things work when you try to set it to -1?  from the jira it seems that was a major goal here, to prevent incorrect results.

**Comment by squito** [comment]
though I wonder if instead it should fail (with a good error) if you give a bad value, rather than just logging a message and ignoring the request.  thought @marmbrus?

**Comment by marmbrus** [comment]
Warnings seem likely to be missed.  I would throw an exception.

**Comment by viirya** [comment]
Thanks. Updated to throw exception now. Also added a test for it.

**Comment by viirya** [comment]
should not be a related failure.

**Comment by viirya** [comment]
test again please.

**Comment by squito** [comment]
lgtm

**Comment by squito** [comment]
Jenkins, retest this please

**Comment by squito** [comment]
lgtm

**Comment by TanninOne** [comment]
Has anyone verified that a **string** comparison against -1 completely fully fixes the issue? Isn't it interpreted as an integer later?
What I'm getting at is: What if the user has the variable to "-01". or to "-2"? Maybe those were valid values for mapred.reduce.tasks as well? How was zero handled in hadoop?
Wouldn't it be safer and cleaner to test any variable against a valid range?
I.e. "if (Integer.parseInt("value") < 1) ..."

**Comment by viirya** [comment]
Setting `mapred.reduce.tasks` to '-1' has special meaning that asks Hive to automatically determine the number of reducers. It is the default value in [hive-default.xml](https://issues.apache.org/jira/browse/HIVE-490). You can also refer to [`HADOOPNUMREDUCERS`](https://github.com/apache/hive/blob/c72d073c1fe2a07c80120647bb2170fb7e50d168/common/src/java/org/apache/hadoop/hive/conf/HiveConf.java#L378).

Since it will parse the given string value, the configuration value of "-01" or "-2" should be able to be parsed to integer value too. As I can tell from the Hive sources, it checks if `mapred.reduce.tasks` is greater than 0 before using it in [SemanticAnalyzer](https://github.com/apache/hive/blob/aaa112dbfd0b52285925dad4160f34487b8dfd82/ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java#L6077) and [SetReducerParallelism](https://github.com/apache/hive/blob/aaa112dbfd0b52285925dad4160f34487b8dfd82/ql/src/java/org/apache/hadoop/hive/ql/optimizer/SetReducerParallelism.java#L66). So I think these values "-01", "-2" would have the same effect as setting it as '-1'.

Although I don't think it will be possible to have such values as setting values for `mapred.reduce.tasks`, it is still good to consider these cases. I will update it later.

**Comment by TanninOne** [comment]
Great, thanks for looking into this.

**Comment by marmbrus** [comment]
Thanks, merged to master and 1.4

**Pull request closed without merging**